File: agent/node/report.ts
import type { Mapping } from "./mapping.ts";
import type { ScopeGuard } from "./scope.ts";
import { plan } from "./mapping.ts";

export type Crumb = {
  authorId: string;
  authorName: string;
  claim: string;
  permalink: string;
};

export function formatCrumb(crumb: Crumb): string | null {
  if (!crumb.permalink || !crumb.claim) return null;
  const who = crumb.authorName || `<unresolved:${crumb.authorId}>`;
  return `  - ${who}: ${crumb.claim}  [${crumb.permalink}]`;
}

export function formatReport(opts: {
  mapping: Mapping;
  guard: ScopeGuard;
  broadcastId: string;
  replyCount: number;
  collectSeconds: number;
  trails?: Record<string, Crumb[]>;
}): string {
  if (opts.mapping.work.length) return formatWorkReport(opts.mapping, opts.guard, opts.trails ?? {});
  opts.guard.require(opts.broadcastId, "report");
  const rooms = opts.mapping.channels.filter((c) => c.role === "room");
  const reports = opts.mapping.channels.filter((c) => c.role === "reports");
  return [
    `# report from ${opts.mapping.agent}`,
    `scope: ${opts.mapping.scope}`,
    `broadcast: ${opts.guard.redact(opts.broadcastId)}`,
    `rooms: ${rooms.length}`,
    `replies: ${opts.replyCount}`,
    `window: ${opts.collectSeconds}s`,
    `filed: ${reports.length} reports channel(s)`,
    "",
  ].join("\n");
}

function formatWorkReport(
  mapping: Mapping,
  guard: ScopeGuard,
  trails: Record<string, Crumb[]>,
): string {
  const rows = plan(mapping);
  const posts = rows.filter((r) => r.action === "post").length;
  const gathers = rows.filter((r) => r.action === "gather").length;
  const skips = rows.filter((r) => r.action === "skip").length;
  const gaps = rows.filter((r) => r.action === "gap").length;
  const lines = [
    `# report from ${mapping.agent}`,
    `scope: ${mapping.scope}`,
    `work: ${rows.length}  post: ${posts}  gather: ${gathers}  skip: ${skips}  gaps: ${gaps}`,
  ];
  for (const { work: w, action } of rows) {
    lines.push("");
    lines.push(`## ${w.id} — ${w.state || "active"}`);
    if (action === "gap") {
      lines.push("gaps: room not in map");
      continue;
    }
    lines.push(`room: ${guard.redact(w.room)}`);
    if (w.owner) lines.push(`owner: ${w.owner}`);
    if (action === "post") {
      lines.push("anchor: (none — post, then gather next run)");
      lines.push("trail: (empty until the thread exists)");
    } else if (action === "skip") {
      lines.push("skip: closed");
    } else {
      lines.push(`anchor: ${w.anchor}`);
      lines.push("action: gather");
      const crumbs = (trails[w.id] ?? []).map(formatCrumb).filter((c): c is string => c !== null);
      if (crumbs.length) {
        lines.push("trail:");
        lines.push(...crumbs);
      } else {
        lines.push("trail: silent (read this thread; do not open another)");
      }
    }
  }
  lines.push("");
  return lines.join("\n");
}

File: agent/python/report.py
"""Broadcast plan and breadcrumb trail. A crumb is a cited reply, or it is omitted."""

from __future__ import annotations

from dataclasses import dataclass

from mapping import Mapping
from scope import ScopeGuard


@dataclass(frozen=True)
class Crumb:
    author_id: str
    author_name: str
    claim: str
    permalink: str


def format_crumb(crumb: Crumb) -> str | None:
    if not crumb.permalink or not crumb.claim:
        return None
    who = crumb.author_name or ("<unresolved:" + crumb.author_id + ">")
    return f"  - {who}: {crumb.claim}  [{crumb.permalink}]"


def format_report(
    *,
    mapping: Mapping,
    guard: ScopeGuard,
    broadcast_id: str,
    reply_count: int,
    collect_seconds: int,
    trails: dict[str, tuple[Crumb, ...]] | None = None,
) -> str:
    if mapping.work:
        return _format_work_report(mapping, guard, trails or {})
    guard.require(broadcast_id, "report")
    rooms = mapping.by_role("room")
    reports = mapping.by_role("reports")
    lines = [
        f"# report from {mapping.agent}",
        f"scope: {mapping.scope}",
        f"broadcast: {guard.redact(broadcast_id)}",
        f"rooms: {len(rooms)}",
        f"replies: {reply_count}",
        f"window: {collect_seconds}s",
        f"filed: {len(reports)} reports channel(s)",
    ]
    return "\n".join(lines) + "\n"


def _format_work_report(
    mapping: Mapping,
    guard: ScopeGuard,
    trails: dict[str, tuple[Crumb, ...]],
) -> str:
    plan = mapping.plan()
    posts = sum(1 for _, a in plan if a == "post")
    gathers = sum(1 for _, a in plan if a == "gather")
    skips = sum(1 for _, a in plan if a == "skip")
    gaps = sum(1 for _, a in plan if a == "gap")
    lines = [
        f"# report from {mapping.agent}",
        f"scope: {mapping.scope}",
        f"work: {len(plan)}  post: {posts}  gather: {gathers}  skip: {skips}  gaps: {gaps}",
    ]
    for w, action in plan:
        lines.append("")
        lines.append("## " + w.id + " — " + (w.state or "active"))
        if action == "gap":
            lines.append("gaps: room not in map")
            continue
        lines.append(f"room: {guard.redact(w.room)}")
        if w.owner:
            lines.append(f"owner: {w.owner}")
        if action == "post":
            lines.append("anchor: (none — post, then gather next run)")
            lines.append("trail: (empty until the thread exists)")
        elif action == "skip":
            lines.append("skip: closed")
        else:
            lines.append(f"anchor: {w.anchor}")
            lines.append("action: gather")
            crumbs = [c for c in (format_crumb(x) for x in trails.get(w.id, ())) if c]
            if crumbs:
                lines.append("trail:")
                lines.extend(crumbs)
            else:
                lines.append("trail: silent (read this thread; do not open another)")
    return "\n".join(lines) + "\n"
